llama/model.py

```diff
@@ -87,100 +95,114 @@ def repeat_kv(x: torch.Tensor, n_rep: int) -> torch.Tensor:
 
 
 class Attention(nn.Module):
-    def __init__(self, args: ModelArgs):
+    def __init__(self, args: ModelArgs, world_size: Optional[int] = None,
+                 rank: Optional[int] = None, groups: Optional[List] = None):
         super().__init__()
-        self.n_kv_heads = args.n_heads if args.n_kv_heads is None else args.n_kv_heads
-        model_parallel_size = fs_init.get_model_parallel_world_size()
-        self.n_local_heads = args.n_heads // model_parallel_size
-        self.n_local_kv_heads = self.n_kv_heads // model_parallel_size
-        self.n_rep = self.n_local_heads // self.n_local_kv_heads
+        #NEW ADDITION
+        #self.n_kv_heads = args.n_heads if args.n_kv_heads is None else args.n_kv_heads
+        #model_parallel_size = fs_init.get_model_parallel_world_size()
+        #self.n_local_heads = args.n_heads // model_parallel_size
+        #self.n_local_kv_heads = self.n_kv_heads // model_parallel_size
+        #self.n_rep = self.n_local_heads // self.n_local_kv_heads
+
+        if world_size is None:
+            groups = get_model_parallel_group()
+            world_size = get_model_parallel_world_size()
+            rank = get_model_parallel_rank()
+
+        self.n_local_heads = divide_and_check_no_remainder(args.n_heads, world_size)
         self.head_dim = args.dim // args.n_heads
 
+        init_method = lambda x: x
+
         self.wq = ColumnParallelLinear(
             args.dim,
             args.n_heads * self.head_dim,
             bias=False,
             gather_output=False,
-            init_method=lambda x: x,
+            init_method=lambda x: x, #THIS CALL IS MADE SIMPLER: WORLD_SIZE, RANK, GROUP
```

Comment:
The `world_size`, `rank`, `group` needs to be passed in from here to allow us arbitrarily instantiate a shard of the model to do resharding. It can not be simplified.

Reply:
yes - all needed parameters were added while cleaning the `model.py` comments.